src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala

```diff
@@ -364,41 +372,79 @@ trait ExprTyping extends BaseTyping { this: TypeInfoImpl =>
     case e => violation(s"unexpected expression $e")
   }
 
-  /** returns a type that is implied by the context if the numeric expression is an untyped constant expression */
-  private def getTypeFromContext(expr: PNumExpression): Option[Type] = expr match {
-    case tree.parent(p) => p match {
-      // if no type is specified, integer constants default to int in short var declarations
-      case _: PShortVarDecl => Some(IntT(config.typeBounds.Int))
-      case PAssignmentWithOp(_, _, pAssignee) => Some(exprType(pAssignee))
-      case PAssignment(rights, lefts) =>
-        val index = rights.indexOf(expr)
-        lefts(index) match {
-          case PBlankIdentifier() =>
-            // if no type is specified, integer constants default to int in assignments to blank identifiers
-            Some(IntT(config.typeBounds.Int))
-          case x => Some(exprType(x))
-        }
-      case PConstDecl(typ, _, _) => typ map typeSymbType
-      // if no type is specified, integer constants default to int in var declarations
-      case PVarDecl(typ, _, _, _) => if(typ.isEmpty) Some(IntT(config.typeBounds.Int)) else typ map typeSymbType
-      case n: PInvoke => resolve(n) match {
-        case Some(ap.FunctionCall(callee, args)) =>
-          val index = args.indexOf(expr)
-          callee match {
-            case f: ap.Function => Some(typeSymbType(f.symb.args(index).typ))
-            case _ => None
+  /** Returns a non-interface type that is implied by the context if the numeric expression is an untyped
+    * constant expression. It ignores interface types. This is useful to check the bounds of constant expressions when
+    * they are assigned to a variable of an interface type. For those cases, we need to obtain the numeric type of the
+    * expression.
+    */
+  private def getNonInterfaceTypeFromCtxt(expr: PNumExpression): Option[Type] = {
+    // if an unbounded integer constant expression is assigned to an interface type,
+    // then it has the default type
+    def defaultTypeIfInterface(t: Type) : Type = {
+      if (t.isInstanceOf[InterfaceT]) DEFAULT_INTEGER_TYPE else t
+    }
```

Comment:
Why not write a function that returns the typeFromContext and then you apply the defaultTypeIfInterface afterwards? Then you need a less specialzed function.